query-connector/src/app/database-service.ts

```diff
@@ -453,3 +452,118 @@ export async function insertQuery(input: QueryInput) {
+    const foundVS = result.rows[0];
+    if (
+      foundVS.version !== vs.valueSetVersion ||
+      foundVS.name !== vs.valueSetName ||
+      foundVS.author !== vs.author
+    ) {
+      console.error(
+        "Retrieved value set information differs from given value set",
+      );
+      missingData.missingValueSet = true;
+    }
+  } catch (error) {
+    console.error("Couldn't fetch inserted value set from DB: ", error);
+    missingData.missingValueSet = true;
+  }
+
+  // Check that all concepts under the value set's umbrella were inserted
+  const brokenConcepts = await Promise.all(
+    vs.concepts.map(async (c) => {
+      const systemPrefix = stripProtocolAndTLDFromSystemUrl(vs.system);
+      const conceptId = `${systemPrefix}_${c.code}`;
+      const conceptSql = `SELECT * FROM concepts WHERE id = $1;`;
+
+      try {
+        const result = await dbClient.query(conceptSql, [conceptId]);
+        const foundConcept: Concept = result.rows[0];
+
+        // We accumulate the unique DIBBs concept IDs of anything that's missing
+        if (
+          foundConcept.code !== c.code ||
+          foundConcept.display !== c.display
+        ) {
+          console.error(
+            "Retrieved concept " +
+              conceptId +
+              " has different values than given concept",
+          );
+          return conceptId;
+        }
+      } catch (error) {
+        console.error(
+          "Couldn't fetch concept with ID " + conceptId + ": ",
+          error,
+        );
+        return conceptId;
+      }
+    }),
+  );
+  missingData.missingConcepts = brokenConcepts.filter((bc) => bc !== undefined);
+
+  // Confirm that valueset_to_concepts contains all relevant FK mappings
+  const mappingSql = `SELECT * FROM valueset_to_concept WHERE valueset_id = $1;`;
+  try {
+    const result = await dbClient.query(mappingSql, [vs.valueSetId]);
+    const rows = result.rows;
+    const missingConceptsFromMappings = vs.concepts.map((c) => {
+      const systemPrefix = stripProtocolAndTLDFromSystemUrl(vs.system);
+      const conceptUniqueId = `${systemPrefix}_${c.code}`;
+
+      // Accumulate unique IDs of any concept we can't find among query rows
+      const fIdx = rows.findIndex((r) => r["concept_id"] === conceptUniqueId);
+      if (fIdx === -1) {
+        console.error(
+          "Couldn't locate concept " + conceptUniqueId + " in fetched mappings",
+        );
+        return conceptUniqueId;
+      }
+    });
+    missingData.missingMappings = missingConceptsFromMappings.filter(
+      (item) => item !== undefined,
+    );
+  } catch (error) {
+    console.error(
+      "Couldn't fetch value set to concept mappings for this valueset: ",
+      error,
+    );
+    const systemPrefix = stripProtocolAndTLDFromSystemUrl(vs.system);
+    vs.concepts.forEach((c) => {
+      const conceptUniqueId = `${systemPrefix}_${c.code}`;
+      missingData.missingMappings.push(conceptUniqueId);
+    });
+  }
+
+  return missingData;
```

Comment:
Are you thinking that we might display this back to the user somehow?

Reply:
I think it could have a lot of different uses. I think display is one of them, but I was actually just imagining it as a data structure for the insert script to use. There are different cases of missingness if an insertion fails, and we need to be able to capture the extent to which things are missing. E.g., we might have successfully inserted all concepts, but not the value set itself, which means we also failed on foreign key mappings. I thought a succinct object that captured each of these dimensions separately was the cleanest way for another part of the code (which I think is a separate helper function when we write the script, something like

```
handleValueSetReInsertion(missingData) {
  if (missingData.missingValueSet) { ... }
  ...
}
```
) to then go back in and take care of whatever we were missing in a graceful and organized way. I imagined this as fairly internal which is why I didn't make an explicit type for it, but I can if we intend to display it.

Reply:
Gotcha! That makes sense to me.